### PR TITLE
Network error

### PR DESCRIPTION
**Pull request by hannesm**
see https://github.com/mirage/mirage-dev/pull/166

### CODE REVIEW

File: pkg/pkg.ml [hunk 1]
```diff
@@ -33,7 +33,10 @@ let () =
     Ok [ Pkg.lib "pkg/META.mirage-types" ~dst:"META";
          Pkg.lib "types/V1.mli";
          Pkg.lib "types/V1.cmi";
+         Pkg.lib "types/V1.cmti";
          Pkg.lib ~cond:lwt "types/V1_LWT.mli";
-         Pkg.lib ~cond:lwt "types/V1_LWT.cmi"; ]
+         Pkg.lib ~cond:lwt "types/V1_LWT.cmi";
+         Pkg.lib ~cond:lwt "types/V1_LWT.cmti";
+         Pkg.mllib "types/mirage-types-pp.mllib"; ]
```
Comment (samoht) [inline pkg/pkg.ml]: this should just be `mirage-types.mllib` and we should start adding useful code in there: pretty printers are a good start!

File: pkg/pkg.ml [hunk 2]
```diff
@@ -33,7 +33,10 @@ let () =
     Ok [ Pkg.lib "pkg/META.mirage-types" ~dst:"META";
          Pkg.lib "types/V1.mli";
          Pkg.lib "types/V1.cmi";
+         Pkg.lib "types/V1.cmti";
          Pkg.lib ~cond:lwt "types/V1_LWT.mli";
-         Pkg.lib ~cond:lwt "types/V1_LWT.cmi"; ]
```
Comment (dbuenzli) [inline pkg/pkg.ml]: Are these mli only compilation units ? I should introduce an `Exts.types_api` shortcut for these.

Comment (dbuenzli) [inline pkg/pkg.ml]: See https://github.com/dbuenzli/topkg/issues/74

Comment (samoht) [inline pkg/pkg.ml]: yes and I remember doing similar things in other packages too

File: pkg/pkg.ml (hunk 1 above)
Comment (hannesm) [inline pkg/pkg.ml]: but I couldn't figure out how to put V1 in the mllib without an implementation, and V1_LWT only conditionally...

File: pkg/pkg.ml (hunk 2 above)
Comment (dbuenzli) [inline pkg/pkg.ml]: Aaah @samoht I'm very disappointed you didn't complain.

**Comment by hannesm** [comment]
I updated this PR, now there is a `mirage-types-runtime` (providing atm pretty printers) and a `mirage-types-runtime-lwt` (providing useful combinators for combining result + lwt).

**Comment by hannesm** [comment]
(if there's a way to include `V1.mli` and `M_pp.ml` and `M_pp.mli` (without having to write a `V1.ml`) in a single `mllib` file, please let me know to avoid this library explosion)

File: types/V1.mli [hunk 3]
```diff
@@ -299,38 +302,53 @@ module type BLOCK = sig
 
 end
 
-(** {1 Network stack}
+(** {1 Network stack} *)
+
+module Network : sig
+  type error = [
+    | `Unknown of string (** an undiagnosed error *)
```
Comment (yomimono) [inline types/V1.mli]: Perhaps we should change `Unknown` to `Msg` to fit in more nicely with Rresult?

Comment (hannesm) [inline types/V1.mli]: sure, and imho remove `` `Disconnected`` (see https://lists.xenproject.org/archives/html/mirageos-devel/2016-10/msg00020.html).  In here I intentionally did not rename or remove anything, but `only` put in result and centralised the error polyvar (and pp).

**Comment by samoht** [comment]
Yes there is ;-)

```
$ cat types/mirage-types.mllib
runtime/M_pp
```

```
$ cat pkg/META.types
version = "%%VERSION%%"
description = "Collection of module signatures for MirageOS"
exists_if = "V1.cmi"
requires = "result"
archive(byte) = "mirage-types.cma"
archive(native) = "mirage-types.cmxa"
plugin(byte) = "mirage-types.cma"
plugin(native) = "mirage-types.cmxs"
exists_if = "mirage-types.cma"
```

etc.

The only thing to remember is that`mli/cmi` never appear in META files (because you only need them at compile time, and they are made visible with `-I`) but they need to be installed properly, so the `cmi/cmti` lines that you added in `pkg/pkg.ml` are indeed needed.

**Comment by hannesm** [comment]
@samoht thanks, so no mention of V1 in the mllib?

**Comment by hannesm** [comment]
force-pushed thx to @samoht (also force-pushed my tcpip branch)

**Comment by samoht** [comment]
exactly, don't mention the mli files in the mllib/META.

File: types/V1.mli [hunk 4]
```diff
@@ -299,38 +302,53 @@ module type BLOCK = sig
...
   (** [write nf buf] outputs [buf] to netfront [nf]. *)
 
-  val writev: t -> buffer list -> unit io
+  val writev: t -> buffer list -> (unit, error) result io
   (** [writev nf bufs] output a list of buffers to netfront [nf] as a
       single packet. *)
 
-  val listen: t -> (buffer -> unit io) -> unit io
+  (* XXX: listen currently runs forever (documentation wrong)! *)
+  val listen: t -> (buffer -> unit io) -> (unit, error) result io
```
Comment (yomimono) [inline types/V1.mli]: This `buffer -> unit io` means the function we call can't pass us any errors -- we can only emit those we caused ourselves.  When attempting to revise users to use result types, this presents some difficulties.  Is it intentional, or an opportunity for more discussion?

Comment (talex5) [inline types/V1.mli]: (also, the second `unit` should be `'a` if it never returns except on error)

**Comment by hannesm** [comment]
I rebased on master, now use `%%VERSION_NUM%%` in META, and got again rid of the lwt result combinators since [Lwt_result.Infix](https://ocsigen.org/lwt/dev/api/Lwt_result.Infix) exists (plus discussion about naming on the mailing list).  From my perspective this is ready to merge (as said in the Mirage call today), and we can iterate over the API in subsequent commits (as discussed here, rename `` `Unknown of string `` to `` `Msg of string ``, find a reasonable signature for `listen`, etc.).

**Comment by hannesm** [comment]
I restarted the mirage-dev build over [here](https://travis-ci.org/mirage/mirage-dev/builds/165685332) after the last rebase + force-push

**Comment by hannesm** [comment]
which just now turned all green...

**Comment by yomimono** [comment]
In the future I'll make sure to mention that I'm basing my own revisions off of your branches, in hopes of avoiding having to clean up after force-pushes.

I'll merge to make this less messy and because having better error handling for the individual `NETWORK` implementations has some value outside of the higher levels of the TCP/IP stack.  I expect this API (particularly for `listen`) to change further, as @hannesm mentioned.

**Comment by hannesm** [comment]
thx, agreed!  @yomimono sorry for rebasing and force-pushing so often...

**Pull request merged**